tokio/src/sync/semaphore.rs

```diff
@@ -123,6 +123,117 @@ use std::sync::Arc;
 /// # }
 /// ```
 ///
+/// ## Prevent tests from running in parallel
+///
+/// By default, Rust runs tests in the same file in parallel. However, in some cases, running two tests in parallel may lead to problems.
+/// For example, this can happen when tests use the same database.
+///
+/// Consider the following scenario:
+/// 1. `test_insert`: Inserts a key-value pair into the database, then retrieves the value using the same key to verify the insertion.
+/// 2. `test_update`: Inserts a key, then updates the key to a new value and verifies that the value has been accurately updated.
+/// 3. `test_others`: A third test that doesn't modify the database state. It can run in parallel with the other tests.
+///
+/// In this example, `test_insert` and `test_update` need to run in sequence to work, but it doesn't matter which test runs first.
+/// We can leverage a semaphore with a single permit to address this challenge.
+///
+/// ```
+/// use tokio::sync::{Semaphore, Mutex};
+/// use lazy_static::lazy_static;
+/// # use std::collections::HashMap;
+/// # struct Database {
+/// #   map: HashMap<String, i32>,
+/// # }
+/// # impl Database {
+/// #    pub fn setup() -> Database {
+/// #        Database {
+/// #            map: HashMap::new(),
+/// #        }
+/// #    }
+/// #    pub fn insert(&mut self, key: &str, value: i32) {
+/// #        self.map.insert(key.to_string(), value);
+/// #    }
+/// #    pub fn update(&mut self, key: &str, value: i32) {
+/// #        self.map
+/// #            .entry(key.to_string())
+/// #            .and_modify(|origin| *origin = value);
+/// #    }
+/// #    pub fn delete(&mut self, key: &str) {
+/// #        self.map.remove(key);
+/// #    }
+/// #    pub fn get(&mut self, key: &str) -> i32 {
+/// #        *self.map.get(key).unwrap()
+/// #    }
+/// # }
+/// // Initialize a static semaphore with only one permit, which is used to
+/// // prevent test_insert and test_update from running in parallel.
+/// static PERMIT: Semaphore = Semaphore::const_new(1);
+///
+/// // Initialize the database that will be used by the subsequent tests.
+/// lazy_static! {
+///     static ref DB: Mutex<Database> = Mutex::new(Database::setup());
+/// }
```

Comment:
Since `Mutex::new` is const, you don't need `lazy_static!` for this.
```suggestion
/// static DB: Mutex<Database> = Mutex::new(Database::setup());
```

Reply:
I am in a dilemma.
In code 1, it needs something like once_cell / Lazy to pass compiler check.
In code 2, It will expose the Mutex to reader,  we want avoid it.


code 1 
```
struct Database {
    map: Mutex<HashMap<String, i32>>,
}
pub fn setup() -> Database {
    Database {
        map: Mutex::new(HashMap::new()),
    }
}
static DB: Database = Database::setup();  
```
compiler complains that: 
> error[E0015]: cannot call non-const fn `Database::setup` in statics
  --> src/main.rs:36:23
   |
36 | static DB: Database = Database::setup();
   |                       ^^^^^^^^^^^^^^^^^
   |
   = note: calls in statics are limited to constant functions, tuple structs and tuple variants
   = note: consider wrapping this expression in `Lazy::new(|| ...)` from the `once_cell` crate: https://crates.io/crates/once_cell

code 2, 
```
struct Database {
    map: HashMap<String, i32>,
}
static DB: Mutex<Database> = Mutex::new(Database::setup());
```

Reply:
Hm, I think the problem is the hash map. It's constructor is not const. Try this instead:
```
struct Database {
    map: Mutex<BTreeMap<String, i32>>,
}
pub const fn setup() -> Database {
    Database {
        map: Mutex::new(BTreeMap::new()),
    }
}
static DB: Database = Database::setup();  
```
Here I replaced `HashMap` with `BTreeMap` and added `const` to `setup`.

Reply:
Fixed.